prometheus/src/main/java/org/opensearch/sql/prometheus/storage/implementor/PrometheusDefaultImplementor.java

```diff
@@ -31,26 +33,86 @@
 public class PrometheusDefaultImplementor
     extends DefaultImplementor<PrometheusMetricScan> {
 
+
+  @Override
+  public PhysicalPlan visitNode(LogicalPlan plan, PrometheusMetricScan context) {
+    if (plan instanceof PrometheusLogicalMetricScan) {
+      return visitIndexScan((PrometheusLogicalMetricScan) plan, context);
+    } else if (plan instanceof PrometheusLogicalMetricAgg) {
+      return visitIndexAggregation((PrometheusLogicalMetricAgg) plan, context);
+    } else {
+      throw new IllegalStateException(StringUtils.format("unexpected plan node type %s",
+          plan.getClass()));
+    }
+  }
+
+  /**
+   * Implement PrometheusLogicalMetricScan.
+   */
+  public PhysicalPlan visitIndexScan(PrometheusLogicalMetricScan node,
+                                     PrometheusMetricScan context) {
+    SeriesSelectionQueryBuilder seriesSelectionQueryBuilder = new SeriesSelectionQueryBuilder();
+    String query = seriesSelectionQueryBuilder.build(node.getMetricName(), node.getFilter());
+
+    context.getRequest().setPromQl(query);
+    setTimeRangeParameters(node.getFilter(), context);
+    context.getRequest()
+        .setStep(new StepParameterResolver().resolve(context.getRequest().getStartTime(),
+            context.getRequest().getEndTime(), null));
+    return context;
+  }
+
+  /**
+   * Implement PrometheusLogicalMetricAgg.
+   */
+  public PhysicalPlan visitIndexAggregation(PrometheusLogicalMetricAgg node,
+                                            PrometheusMetricScan context) {
+
+    setTimeRangeParameters(node.getFilter(), context);
+    context.getRequest()
+        .setStep(new StepParameterResolver().resolve(context.getRequest().getStartTime(),
+            context.getRequest().getEndTime(), node.getGroupByList()));
+    String step = context.getRequest().getStep();
+    SeriesSelectionQueryBuilder seriesSelectionQueryBuilder = new SeriesSelectionQueryBuilder();
+    String seriesSelectionQuery
+        = seriesSelectionQueryBuilder.build(node.getMetricName(), node.getFilter());
+
+    AggregationQueryBuilder aggregationQueryBuilder = new AggregationQueryBuilder();
+    String aggregateQuery
+        = aggregationQueryBuilder.build(node.getAggregatorList(),
+        node.getGroupByList());
+
+    String finalQuery = String.format(aggregateQuery, seriesSelectionQuery + "[" + step + "]");
```

Comment:
You may want add high level abstraction or more comments in future. How to compose finalQuery is the responsebility of QueryBuilder. QueryBuilder is compose of [aggQuery, parameters] and implemement String build() interface.

Reply:
Will refactor after 2.4 release.